libraries/streaming-e2e-tests/react-app/src/App.js

```diff
@@ -1,12 +1,84 @@
-import React from 'react';
+import React, { Fragment, useCallback, useEffect, useMemo, useReducer } from 'react';
 import ReactWebChat from 'botframework-webchat';
+import { DirectLineStreaming } from 'botframework-directlinejs';
 
-function App({directLine}) {
-  // nonce related to botbuilder-js content security issue #2762:
-  // https://github.com/microsoft/botbuilder-js/issues/2762
-  return (
-    <ReactWebChat directLine={directLine} nonce="a1b2c3d" />
-  );
+const actions = {
+    error: 'error',
+    token: 'token',
+};
+
+function App({ bot }) {
+    if (!bot) {
+        throw new Error('bot prop is required');
+    }
+
+    const hostname = `https://${bot}.azurewebsites.net`;
+
+    const [state, dispatch] = useReducer(
+        (state, action) => {
+            switch (action.type) {
+                case actions.error:
+                    return { ...state, error: action.value, token: null };
+
+                case actions.token:
+                    return { ...state, error: null, token: action.value };
+
+                default:
+                    return state;
+            }
+        },
+        { error: null, token: null }
+    );
+
+    // Synchronize token with hostname
+    useEffect(() => {
+        (async () => {
```

Comment:
Use `AbortController` to stop the async function if the component unmount sooner.

```js
useEffect(() => {
  const abortController = new AbortController();

  (async function () {
    const res = await fetch('...', { signal: abortController.signal });

    ...

    const { token } = await res.json();

    if (abortController.signal.aborted) {
      return;
    }

    dispatch({ ..., payload: token });
  })();

  return () => abortController.abort();
}, [...]);
```

Reply:
That's slick! I used to use the `let mounted = true, return () => mounted = false` and conditionally update state if mounted was `true`. Much cleaner.